🧪⇪ Bump `galaxy_ng` to v4.9.2 in tests

webknjaz opened the pull request:
It updates the underlying `pulp_ansible` to v0.20.2, which fixes a race condition that is occasionally observable in our CI when provisioning the initial dummy collections for testing.

Refs:

  * https://github.com/ansible/galaxy_ng/releases/tag/4.9.0
  * https://github.com/ansible/galaxy_ng/pull/1950
  * https://github.com/ansible/galaxy/issues/3267
  * https://github.com/pulp/pulp_ansible/issues/1571
  * https://github.com/pulp/pulp_ansible/pull/1572
  * https://github.com/pulp/pulp_ansible/pull/1595
  * https://github.com/pulp/pulp_ansible/issues/1623
  * https://github.com/pulp/pulp_ansible/pull/1624

##### SUMMARY

$sbj.

##### ISSUE TYPE

- Maintenance Pull Request
- Test Pull Request

##### ADDITIONAL INFORMATION

N/A

webknjaz committed:
🧪⇪ Bump `galaxy_ng` to v4.9.0 in tests

It updates the underlying `pulp_ansible` to v0.20.2, which fixes a
race condition that is occasionally observable in our CI when
provisioning the initial dummy collections for testing.

Refs:

  * https://github.com/ansible/galaxy_ng/releases/tag/4.9.0
  * https://github.com/ansible/galaxy_ng/pull/1950
  * https://github.com/ansible/galaxy/issues/3267
  * https://github.com/pulp/pulp_ansible/issues/1571
  * https://github.com/pulp/pulp_ansible/pull/1572
  * https://github.com/pulp/pulp_ansible/pull/1595
  * https://github.com/pulp/pulp_ansible/issues/1623
  * https://github.com/pulp/pulp_ansible/pull/1624

webknjaz commented:
This is what's broken in the CI runs:
```console
03:20 TASK [ansible-galaxy-collection : ensure we can download the published collection - pulp_v2] ***
03:21 fatal: [testhost]: FAILED! => {"changed": true, "cmd": ["ansible-galaxy", "collection", "install", "-s", "pulp_v2", "-p", "/tmp/ansible.mk27mu05.test/publish/pulp_v2", "ansible_test.my_collection==1.0.0", "-v"], "delta": "0:00:00.470485", "end": "2023-12-11 13:56:52.273680", "msg": "non-zero return code", "rc": 1, "start": "2023-12-11 13:56:51.803195", "stderr": "\u001b[1;35m[WARNING]: The specified \"api_version\" configuration for the galaxy server\u001b[0m\n\u001b[1;35m\"pulp_v2\" is not a public configuration, and may be removed at any time without\u001b[0m\n\u001b[1;35mwarning.\u001b[0m\n\u001b[1;35m[WARNING]: The specified collections path\u001b[0m\n\u001b[1;35m'/tmp/ansible.mk27mu05.test/publish/pulp_v2' is not part of the configured\u001b[0m\n\u001b[1;35mAnsible collections paths\u001b[0m\n\u001b[1;35m'/root/.ansible/collections:/usr/share/ansible/collections'. The installed\u001b[0m\n\u001b[1;35mcollection will not be picked up in an Ansible run, unless within a playbook-\u001b[0m\n\u001b[1;35madjacent collections directory.\u001b[0m\n\u001b[0;31mERROR! Error when getting the collection info for ansible_test.my_collection from pulp_v2 (http://galaxy-pulp/pulp_ansible/galaxy/primary/api/) (HTTP Code: 500, Message: Internal Server Error Code: Unknown)\u001b[0m", "stderr_lines": ["\u001b[1;35m[WARNING]: The specified \"api_version\" configuration for the galaxy server\u001b[0m", "\u001b[1;35m\"pulp_v2\" is not a public configuration, and may be removed at any time without\u001b[0m", "\u001b[1;35mwarning.\u001b[0m", "\u001b[1;35m[WARNING]: The specified collections path\u001b[0m", "\u001b[1;35m'/tmp/ansible.mk27mu05.test/publish/pulp_v2' is not part of the configured\u001b[0m", "\u001b[1;35mAnsible collections paths\u001b[0m", "\u001b[1;35m'/root/.ansible/collections:/usr/share/ansible/collections'. The installed\u001b[0m", "\u001b[1;35mcollection will not be picked up in an Ansible run, unless within a playbook-\u001b[0m", "\u001b[1;35madjacent collections directory.\u001b[0m", "\u001b[0;31mERROR! Error when getting the collection info for ansible_test.my_collection from pulp_v2 (http://galaxy-pulp/pulp_ansible/galaxy/primary/api/) (HTTP Code: 500, Message: Internal Server Error Code: Unknown)\u001b[0m"], "stdout": "\u001b[0;34mUsing /tmp/ansible.mk27mu05.test/galaxy/ansible.cfg as config file\u001b[0m\nStarting galaxy collection install process\nProcess install dependency map", "stdout_lines": ["\u001b[0;34mUsing /tmp/ansible.mk27mu05.test/galaxy/ansible.cfg as config file\u001b[0m", "Starting galaxy collection install process", "Process install dependency map"]}
```

The unescaped stderr:
```console
[WARNING]: The specified "api_version" configuration for the galaxy server
"pulp_v2" is not a public configuration, and may be removed at any time without
warning.
[WARNING]: The specified collections path
'/tmp/ansible.mk27mu05.test/publish/pulp_v2' is not part of the configured
Ansible collections paths
'/root/.ansible/collections:/usr/share/ansible/collections'. The installed
collection will not be picked up in an Ansible run, unless within a playbook-
adjacent collections directory.
ERROR! Error when getting the collection info for ansible_test.my_collection from pulp_v2 (http://galaxy-pulp/pulp_ansible/galaxy/primary/api/) (HTTP Code: 500, Message: Internal Server Error Code: Unknown)
```

webknjaz commented:
Yeah, that's what I thought.

webknjaz commented:
I attached and saw it streaming things like
```python-traceback
pulp [a0feea53400f4988b47590376543557d]: django.request:ERROR: Internal Server Error: /api/galaxy/v3/plugin/ansible/content/primary/collections/index/
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/django/core/handlers/exception.py", line 55, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.8/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/local/lib/python3.8/site-packages/django/views/decorators/csrf.py", line 56, in wrapper_view
    return view_func(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/rest_framework/viewsets.py", line 125, in view
    return self.dispatch(request, *args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/rest_framework/views.py", line 509, in dispatch
    response = self.handle_exception(exc)
  File "/usr/local/lib/python3.8/site-packages/rest_framework/views.py", line 469, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/usr/local/lib/python3.8/site-packages/rest_framework/views.py", line 480, in raise_uncaught_exception
    raise exc
  File "/usr/local/lib/python3.8/site-packages/rest_framework/views.py", line 497, in dispatch
    self.initial(request, *args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/rest_framework/views.py", line 415, in initial
    self.check_permissions(request)
  File "/usr/local/lib/python3.8/site-packages/rest_framework/views.py", line 332, in check_permissions
    if not permission.has_permission(request, self):
  File "/usr/local/lib/python3.8/site-packages/rest_access_policy/access_policy.py", line 69, in has_permission
    allowed = self._evaluate_statements(statements, request, view, action)
  File "/usr/local/lib/python3.8/site-packages/rest_access_policy/access_policy.py", line 113, in _evaluate_statements
    matched = self._get_statements_matching_conditions(
  File "/usr/local/lib/python3.8/site-packages/rest_access_policy/access_policy.py", line 262, in _get_statements_matching_conditions
    passed = self._check_condition(condition, request, view, action)
  File "/usr/local/lib/python3.8/site-packages/rest_access_policy/access_policy.py", line 288, in _check_condition
    result = method(request, view, action)
  File "/usr/local/lib/python3.8/site-packages/galaxy_ng/app/access_control/access_policy.py", line 253, in v3_can_view_repo_content
    distro = ansible_models.AnsibleDistribution.objects.get(base_path=path)
  File "/usr/local/lib/python3.8/site-packages/django/db/models/manager.py", line 87, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/django/db/models/query.py", line 637, in get
    raise self.model.DoesNotExist(
pulp_ansible.app.models.AnsibleDistribution.DoesNotExist: AnsibleDistribution matching query does not exist.
pulp [a0feea53400f4988b47590376543557d]: 127.0.0.1 - admin [11/Dec/2023:18:14:34 +0000] "GET /api/galaxy/v3/plugin/ansible/content/primary/collections/index/?limit=100&offset=0 HTTP/1.0" 500 145 "-" "ansible-httpget"
pulp [f01fb13dcfa242b4b291212793ad06b6]: django.request:ERROR: Internal Server Error: /api/galaxy/v3/plugin/ansible/content/secondary/collections/index/
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/django/core/handlers/exception.py", line 55, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.8/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/local/lib/python3.8/site-packages/django/views/decorators/csrf.py", line 56, in wrapper_view
    return view_func(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/rest_framework/viewsets.py", line 125, in view
    return self.dispatch(request, *args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/rest_framework/views.py", line 509, in dispatch
    response = self.handle_exception(exc)
  File "/usr/local/lib/python3.8/site-packages/rest_framework/views.py", line 469, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/usr/local/lib/python3.8/site-packages/rest_framework/views.py", line 480, in raise_uncaught_exception
    raise exc
  File "/usr/local/lib/python3.8/site-packages/rest_framework/views.py", line 497, in dispatch
    self.initial(request, *args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/rest_framework/views.py", line 415, in initial
    self.check_permissions(request)
  File "/usr/local/lib/python3.8/site-packages/rest_framework/views.py", line 332, in check_permissions
    if not permission.has_permission(request, self):
  File "/usr/local/lib/python3.8/site-packages/rest_access_policy/access_policy.py", line 69, in has_permission
    allowed = self._evaluate_statements(statements, request, view, action)
  File "/usr/local/lib/python3.8/site-packages/rest_access_policy/access_policy.py", line 113, in _evaluate_statements
    matched = self._get_statements_matching_conditions(
  File "/usr/local/lib/python3.8/site-packages/rest_access_policy/access_policy.py", line 262, in _get_statements_matching_conditions
    passed = self._check_condition(condition, request, view, action)
  File "/usr/local/lib/python3.8/site-packages/rest_access_policy/access_policy.py", line 288, in _check_condition
    result = method(request, view, action)
  File "/usr/local/lib/python3.8/site-packages/galaxy_ng/app/access_control/access_policy.py", line 253, in v3_can_view_repo_content
    distro = ansible_models.AnsibleDistribution.objects.get(base_path=path)
  File "/usr/local/lib/python3.8/site-packages/django/db/models/manager.py", line 87, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/django/db/models/query.py", line 637, in get
    raise self.model.DoesNotExist(
pulp_ansible.app.models.AnsibleDistribution.DoesNotExist: AnsibleDistribution matching query does not exist.
```
even before that download test...

sivel commented:
Some of those may be expected. See:

https://github.com/ansible/ansible/blob/a9919dd7f62c9efe17b8acaebf7c627606ae9f66/test/integration/targets/ansible-galaxy-collection/library/reset_pulp.py#L142-L149

webknjaz commented:
jeez

webknjaz commented:
I think this is the relevant bit
```python-traceback
  File "/usr/local/lib/python3.8/site-packages/pulp_ansible/app/galaxy/serializers.py", line 116, in get_latest_version
    "version": rv.version,
AttributeError: 'NoneType' object has no attribute 'version'
```
which suggests another bug in `pulp_ansible` again.

s-hertel commented on test/lib/ansible_test/_internal/commands/integration/cloud/galaxy.py:
```diff
@@ -98,7 +98,7 @@ def __init__(self, args: IntegrationConfig) -> None:
         self.image = os.environ.get(
             'ANSIBLE_PULP_CONTAINER',
-            'quay.io/pulp/galaxy:4.7.1'
+            'quay.io/pulp/galaxy:4.9.0'  # https://github.com/ansible/galaxy_ng
```
Maybe 4.9.2 now? There are a couple issues to work around, but I have these passing locally.
```suggestion
            'quay.io/pulp/galaxy:4.9.2'  # https://github.com/ansible/galaxy_ng
```
It looks like v3 no longer complains if the same artifact is uploaded multiple times, so publish.yml needs to be adjusted, maybe something like:

```diff
diff --git a/test/integration/targets/ansible-galaxy-collection/tasks/publish.yml b/test/integration/targets/ansible-galaxy-collection/tasks/publish.yml
index 1be16ae901..12bd54d8b8 100644
--- a/test/integration/targets/ansible-galaxy-collection/tasks/publish.yml
+++ b/test/integration/targets/ansible-galaxy-collection/tasks/publish.yml
@@ -31,6 +31,21 @@
     chdir: '{{ galaxy_dir }}'
   register: fail_publish_existing
   failed_when: fail_publish_existing is not failed
+  ignore_errors: true
+
+- name: get result of publish collection - {{ test_name }}
+  uri:
+    url: '{{ test_api_server }}v3/plugin/ansible/content/primary/collections/index/ansible_test/my_collection/versions/1.0.0/'
+    return_content: yes
+    user: '{{ pulp_user }}'
+    password: '{{ pulp_password }}'
+    force_basic_auth: true
+  register: publish_collection_update
+  when: fail_publish_existing is not failed
+
+- assert:
+    that:
+      - fail_publish_existing is failed or publish_collection_update["json"]["updated_at"] == publish_collection_actual["json"]["updated_at"]
```

https://github.com/ansible/ansible/pull/82396#issuecomment-1850309920 - the v2 metadata for a collection (/api/v2/collections/namespace/collection_name/) is broken. The versions URL works though. Since v2 has been removed in more recent versions, I'm guessing it won't be fixed, but since we only use this URL to handle refreshing the cache, it is easy to work around:

```diff
diff --git a/lib/ansible/galaxy/api.py b/lib/ansible/galaxy/api.py
index 96991ec365..ab251f9f7b 100644
--- a/lib/ansible/galaxy/api.py
+++ b/lib/ansible/galaxy/api.py
@@ -853,10 +853,18 @@ class GalaxyAPI:
             try:
                 modified_date = self.get_collection_metadata(namespace, name).modified_str
             except GalaxyError as err:
-                if err.http_code != 404:
+                if err.http_code == 404:
+                    # No collection found, return an empty list to keep things consistent with the various APIs
+                    return []
+                elif err.http_code != 500 or 'v3' in self.available_api_versions:
                     raise
-                # No collection found, return an empty list to keep things consistent with the various APIs
-                return []
+
+                # v2 appears to have a bug on some versions, where the metadata URL is an HTTP error 500
+                # since this URL is only used to handle the cache, proceed as if the cache needs to be refreshed.
+                modified_date = None
+                if '%s.%s' % (namespace, name) in modified_cache:
+                    del modified_cache['%s.%s' % (namespace, name)]
+                    self._set_cache()
 
             cached_modified_date = modified_cache.get('%s.%s' % (namespace, name), None)
             if cached_modified_date != modified_date:
```

webknjaz replied:
@s-hertel Thanks, I applied all your patches. Marking this thread as resolved.

webknjaz committed:
🧪⇪ Bump `galaxy_ng` to v4.9.2

Co-authored-by: Sloane Hertel <19572925+s-hertel@users.noreply.github.com>

s-hertel committed:
Work around broken v2 metadata endpoint in Pulp

This patch adjusts the metadata processing to swallow 404 and handle
other cases as a cache reset request.

Ref: https://github.com/ansible/ansible/pull/82396#discussion_r1697523264

s-hertel committed:
Update galaxy publish test for non-erroring Pulp

s-hertel commented on lib/ansible/galaxy/api.py:
```diff
@@ -853,10 +853,18 @@ def get_collection_versions(self, namespace, name):
-                if err.http_code != 404:
+                if err.http_code == 404:
+                    # No collection found, return an empty list to keep things consistent with the various APIs
+                    return []
+                elif err.http_code != 500 or 'v3' in self.available_api_versions:
                     raise
-                # No collection found, return an empty list to keep things consistent with the various APIs
-                return []
+
+                # v2 appears to have a bug on some versions, where the metadata URL is an HTTP error 500
+                # since this URL is only used to handle the cache, proceed as if the cache needs to be refreshed.
+                modified_date = None
+                if '%s.%s' % (namespace, name) in modified_cache:
+                    del modified_cache['%s.%s' % (namespace, name)]
+                    self._set_cache()
```
@webknjaz  Since v2 GalaxyAPI is deprecated and close to removal, I'm second-guessing this. Can we just remove the v2 tests early, instead of adding code to support the lack of fully functional v2 in 4.9.2?

Or, could we keep using an older galaxy ng for v2 coverage  until 2.20 development starts and we can delete everything v2?

webknjaz replied:
@s-hertel I think that coverage is a good reason to keep this. But also, I was looking forward to backporting this to reduce the CI flakiness. I feel like this might be okay to backport, but removals wouldn't be.

Should we merge+backport this and then delete whatever in follow-ups when 2.20 comes? The main value in this for me is the CI stability, which spans across several branches IIRC.

cc @mattclay would this be acceptable to backport for the sake of CI?

mattclay replied:
Since this is a bug fix for V2, it seems like a good idea to backport it. We should continue testing V2 for versions that support it. Once we branch for 2.20 we'll be able to rip out all the V2 code and tests.

s-hertel replied:
If v2 was used by anything other than old galaxy and custom servers mimicking the v2 endpoint, I'd be more in favor of backporting this. This seems featureish for 3rd party galaxy servers, which should conform with https://old-galaxy.ansible.com v2 or GalaxyNG v3.

IMO we should not be adding more code for v2, unless it fixes issues with  https://old-galaxy.ansible.com.

webknjaz replied:
@s-hertel doesn't it fix issues with our own CI where it uses v2 in old branches? How else would we address that?

s-hertel replied:
@webknjaz How will we know fixes on old branches  are compatible with https://old-galaxy.ansible.com/ if we no longer test against a compatible v2? Specifying the api version is internal, so if someone's using a recent version of GalaxyNG, they'll be using v3 via the public API, not v2. We should keep backwards compat with old galaxy since that's how people use v2.

> Or, could we keep using an older galaxy ng for v2 coverage until 2.20 development starts and we can delete everything v2?

Have you taken a look at which GalaxyNG version it actually breaks in and why? Can the image can be updated to a lesser extent to pull in some of the linked fixes and still work with v2?

webknjaz replied:
Dunno, I haven't looked deeper. I was originally hoping for just an image update, but these changes grew on top, as you and sivel suggested.

mattclay reviewed (commented):
I think this looks good overall. Since we're changing code, not just tests, this should be turned into a bugfix and have a changelog fragment.

mattclay commented:
@webknjaz Have you confirmed that the issue this is trying to fix is still occurring in the devel branch? https://github.com/ansible/ansible/pull/84763 included a work-around for one of the galaxy issues we were seeing while working on DT.

webknjaz commented:
@mattclay no, I haven't. I guess I missed that PR. It looks like it might've fixed it. I'll keep an eye out for this test failures.
Are you going to backport that into older branches?

webknjaz commented:
@mattclay I think this is an example of said issue on `stable-2.16`: https://dev.azure.com/ansible/ansible/_build/results?buildId=140568&view=logs&j=70eeb688-fc21-5491-4420-b60b604f62ec&t=d6f17d19-0bb3-5e63-8427-e2ae23c55414.

mattclay commented:
@webknjaz I wasn't planning on backporting the entire PR, but it makes sense to backport the galaxy test fixes. Can you open backports for just those fixes?

webknjaz commented:
Looks like `devel` still has problems with these tests: https://dev.azure.com/ansible/ansible/_build/results?buildId=141717&view=logs&j=02bf57c0-d919-5e2d-1c94-7b4ef705afdf&t=a9b57ab5-2f0b-5a77-0352-06fbabf684da&l=5867 (it's from @s-hertel's PR #84911). Should we still consider this PR, then?

webknjaz commented:
`stable-2.16`: https://dev.azure.com/ansible/ansible/_build/results?buildId=142510&view=logs&j=70eeb688-fc21-5491-4420-b60b604f62ec&t=d6f17d19-0bb3-5e63-8427-e2ae23c55414&l=5231

webknjaz commented:
`stable-2.16`: https://dev.azure.com/ansible/ansible/_build/results?buildId=142998&view=logs&j=114ad447-3cf5-5c3e-d6a4-0cd0b6d5c2c7&s=d65b03ed-7a7b-54a6-b1eb-60fdb469012e&t=04c2859f-ee16-5564-2ccc-f6dee4dfc978&l=5235

`devel`: https://dev.azure.com/ansible/ansible/_build/results?buildId=142999&view=logs&j=02bf57c0-d919-5e2d-1c94-7b4ef705afdf&t=a9b57ab5-2f0b-5a77-0352-06fbabf684da&l=5606 / https://dev.azure.com/ansible/ansible/_build/results?buildId=142999&view=logs&j=70eeb688-fc21-5491-4420-b60b604f62ec&t=d6f17d19-0bb3-5e63-8427-e2ae23c55414&l=5607

@mattclay this makes me think that #84763 didn't fully address the issue and merging DT made it more frequent (two jobs in the same CI run failing like this used to be rare).

mattclay commented:
@webknjaz Thanks for monitoring this issue. Since the random sleep only improves the situation, rather than eliminating the it completely, it occurred to me that we could simply disable the parallel publishing in the test. If the parallel publishing is the source of this issue, then https://github.com/ansible/ansible/pull/85000 should resolve it.

webknjaz commented:
@mattclay thanks! It doesn't seem to have helped: https://github.com/ansible/ansible/pull/85000#issuecomment-2809287835. Pretty sure that it is failing on cleanup, not on publishing.

mattclay commented:
Yeah, this looks like a different error though -- not related to the publishing problems we were seeing.

webknjaz commented:
> Yeah, this looks like a different error though -- not related to the publishing problems we were seeing.

@mattclay perhaps we were talking about different things, then. I checked a few log links in my previous comments above, and all of them exhibit the same race condition in the pulp reset task.

Here's one from today's nightly: https://dev.azure.com/ansible/ansible/_build/results?buildId=143192&view=logs&j=02bf57c0-d919-5e2d-1c94-7b4ef705afdf&t=a9b57ab5-2f0b-5a77-0352-06fbabf684da&l=5473

mattclay commented:
@webknjaz I think part of the problem is that there are different errors for the initial run and the retry. To make things easier to discuss and troubleshoot, I'm disabling the retry on that integration test: https://github.com/ansible/ansible/pull/85013

mattclay commented:
Now that the test isn't retrying, and isn't publishing in parallel, we're now seeing the publishing step time out instead:

https://dev.azure.com/ansible/ansible/_build/results?buildId=143289&view=logs&jobId=70eeb688-fc21-5491-4420-b60b604f62ec&j=70eeb688-fc21-5491-4420-b60b604f62ec&t=d6f17d19-0bb3-5e63-8427-e2ae23c55414

This timeout was occurring previously, but it was more difficult to notice because the test was being retried, and the retry failure was due to concurrency issues -- most likely because the previous publish was still running.

I've opened https://github.com/ansible/ansible/pull/85016 to increase the publish timeout, to see if that's sufficient to get the tests to pass consistently.

webknjaz commented:
> I've opened #85016 to increase the publish timeout, to see if that's sufficient to get the tests to pass consistently.

Looks like that did help. And you're right: I was confusing normal runs vs. reruns.

webknjaz commented:
Here `stable-2.18` hit that `unique_is_highest` bug earlier today:
```json
{
  "rc": 1,
  "stderr": "\u001b[1;35m[WARNING]: The specified \"api_version\" configuration for the galaxy server\u001b[0m\n\u001b[1;35m\"pulp_v2\" is not a public configuration, and may be removed at any time without\u001b[0m\n\u001b[1;35mwarning.\u001b[0m\n\u001b[1;35m[WARNING]: Galaxy import warning message: No changelog found. Add a\u001b[0m\n\u001b[1;35mCHANGELOG.rst, CHANGELOG.md, or changelogs/changelog.yaml file.\u001b[0m\n\u001b[0;31mERROR! Galaxy import process failed: duplicate key value violates unique constraint \"unique_is_highest\"\u001b[0m\n\u001b[0;31mDETAIL:  Key (collection_id, is_highest)=(839aeb95-f5e3-4a2e-a853-1a73a8e4767c, t) already exists.\u001b[0m\n\u001b[0;31m (Code: UNKNOWN)\u001b[0m\n",
  "stdout": "Publishing collection artifact '/root/.ansible/tmp/ansible-moduletmp-1745565311.5552013-pjw5iaof/namespace1-name1-0.0.7/namespace1-name1-0.0.7.tar.gz' to galaxy_ng [http://galaxy-pulp/api/galaxy/content/primary/\nCollection](http://galaxy-pulp/api/galaxy/content/primary//nCollection) has been published to the Galaxy server galaxy_ng [http://galaxy-pulp/api/galaxy/content/primary/\nWaiting](http://galaxy-pulp/api/galaxy/content/primary//nWaiting) until Galaxy import task http://galaxy-pulp/api/galaxy/content/primary/v3/imports/collections/2b6dc0bb-d568-4d8f-bd9e-3d23598fa3ab/ has completed\n"
}
```
*(https://dev.azure.com/ansible/ansible/_build/results?buildId=144050&view=logs&j=70eeb688-fc21-5491-4420-b60b604f62ec&t=d6f17d19-0bb3-5e63-8427-e2ae23c55414&l=1014)*

webknjaz commented:
`stable-2.17`, yesterday:
https://dev.azure.com/ansible/ansible/_build/results?buildId=144220&view=logs&j=70eeb688-fc21-5491-4420-b60b604f62ec&t=d6f17d19-0bb3-5e63-8427-e2ae23c55414&l=1014
